contrib/dict_in_stream/zstd_dict_in_stream_format.md

```diff
@@ -0,0 +1,63 @@
+# Zstandard Dictionary-in-Stream Format
+
+### Version
+
+0.1.0 (2020-10-11): initial version
+
+## Introduction
+
+This document defines a format for including a Zstandard dictionary inside a
+compressed Zstandard stream. When combined with the
+[Seekable Format](../seekable_format) or other formats that use multiple
+Zstandard frames, this format can help reduce the size of each frame without
+requiring an external dictionary. This format can also be used to create a
+stand-alone stream from a stream that uses an external dictionary, without
+needing to recompress the stream.
+
+### Usage
+
+This format is used by tools such as [Wget-AT] to compress WARC files.
+
+[Wget-AT]: https://github.com/ArchiveTeam/wget-lua/releases/tag/v1.20.3-at.20200401.01
+
+## Format
+
+The format consists of a skippable frame containing the dictionary, followed by
+a normal Zstandard stream. All compressed frames in the stream must have been
+compressed using the dictionary.
```

Comment:
No dictionary should also be a valid option for frames.

Reply:
I'm not sure about this. Part of me wants to keep it simple, as an implementation that requires every frame to use the same dictionary may be easier and cleaner than potentially switching between dict and no-dict. On the other hand, other specifications (such as the one for the `.warc.zst` files produced by Wget-AT) can always impose additional restrictions. And I don't see any obvious harm from allowing dictionary-less frames other than that it might lead to the immediate follow-up 'why not allow multiple dictionaries in a stream?', which – while there may well be realistic applications of it – seems like feature creep to me at this point.

Reply:
@JustAnotherArchivist Dictionaries already have an ID and Frames should use that ID if they want to. It doesn't really extend any features, it just registers the dictionaries and frames with their ID can pick it up.

That way you can add several dictionaries, for example for different file types and the encoder can use them as it sees fit. While the improvement will be marginal it gives some flexibility.